Add tests for data_base, push/pop/candidate headers and blocks

kulpreet opened the pull request:
- Restructured push_pop tests and using the v4 data_base methods now
- Still need to add tests for reorganize, invalidate and index public methods
- I ran into NDEBUG some issues with the way ndebug is setup and used.

I am happy for this PR to be merged, pending a review. I can look at the ndebug issue in another PR once we decide if and what changes are required.

### The ndebug issue

Not all configs on travis have --disable-ndebug, and that results in tests having to deal with different error codes returned from methods called. For now, I am using preprocessor directives to check for appropriate returned values or in one case completely skip a test.

We could do one of the two things:

1. Change travis configs to all have --disable-ndebug, or
2. Streamline returned values so that they are the same with or without ndebug being defined
3. We can continue with preprocessor directives as I am doing now

I think we can fix this concern, if needed, in a separate PR.

evoskuil commented:
There’s no relationship between the Travis build configs and error handling requirements.

Use of assertions precludes negative testing that violates the assertions. The presumption when using assertions is that the asserted condition cannot be achieved via public interface. However it’s not always the case that we can sufficiently hide methods using C++. If a method asserts against a condition then that condition cannot be tested and should not be reachable from truly public API.

kulpreet commented:
I see your point that certain conditions in private methods should never occur if the public API is use properly.

I have been writing tests for the private methods and am handling covering error conditions to be sure the verification code is correctly invoked etc. 

I also understand that travis build configs won't affect the error codes on public API, but the private methods return error codes or success depending on ndebug flag.

For example, the push_header method can execute without failure and return success even if the same setup returns an error if ndebug is disabled. For example here: https://github.com/libbitcoin/libbitcoin-database/pull/243/files#diff-0da23093c09be56dd9d67d255d71b15cR536

I could take out the tests that test the verification invocations. However, the process of writing those did help me track a small issue (https://github.com/libbitcoin/libbitcoin-database/pull/243/files#diff-4fa09e1952e40e0ffc98625e7e0a4390L575).

Let me know if you want me to take out the tests that capture verification conditions, or any other changes I should make before you can merge this.

I also added the `confirm(block_hash, height)` to this PR. The test code for it was dependent on this PR anyway. Hope that is OK. https://github.com/libbitcoin/libbitcoin-database/pull/243/commits/4ebbdc8ff8412a80bf302859c98e5e29027a7f4f

kulpreet commented:
> In this case you must made the test ndebug conditional. 

Super. That is what I have done for now in all tests where this is applicable. So I think we are good to review/merge this.

kulpreet commented:
Also, a small note that the new `confirm(block_hash, height)` method is in this PR as well.

evoskuil commented on test/data_base.cpp:
```diff
@@ -1,401 +1,1148 @@
+   BOOST_REQUIRE(instance.create(genesis));
+   
+   const auto block1 = read_block(MAINNET_BLOCK1);
+   auto out_headers = std::make_shared<header_const_ptr_list>(header_const_ptr_list{});
```
Is the empty list parameter required?

kulpreet replied:
Empty list parameter wasn't required. Removed.

evoskuil commented on test/data_base.cpp:
```diff
@@ -1,401 +1,1148 @@
+               });
+      
+   BOOST_REQUIRE(instance.push_all(headers_push_ptr, config::checkpoint(genesis.hash(), 0)));
+   for(const auto block_ptr: *blocks_push_ptr) {
```
Style: whitespace (“for” is not a fn), balance braces.

kulpreet replied:
Fixed everywhere.

kulpreet commented on src/data_base.cpp:
```diff
@@ -323,6 +323,21 @@ code data_base::reorganize(const config::checkpoint& fork_point,
+    
+    // confirm
+    if (!blocks_->index(block_hash, height, false))
+        return error::operation_failed;
```
I should be confirming transactions too. I realise that now.

I thought we don't need to call update on the block to associate the transactions. This is because the block is already candidated, the transactions will be associated properly with the block before we try to confirm the block. Am I right on this?

Am I missing something else?

kulpreet replied:
New commit adding transaction confirmation for the block being confirmed. Test verifies the confirmation.

evoskuil commented:
Yes/no

kulpreet commented on test/data_base.cpp:
```diff
@@ -1,401 +1,1148 @@
+   data_base instance(settings); 
+   
+   static const auto bc_settings = bc::settings(bc::config::settings::mainnet);
+   const chain::block genesis = bc_settings.genesis_block;
```
`auto&` doesn't work because `bc::settings#genesis_block` returns a `config::block`, so we need to cast it to `chain::block`.

Changed all such occurrences to to `const chain::block& genesis = bc_settings.genesis_block;`. This will avoid the extraneous call to the constructor for `chain::block`

kulpreet commented on test/data_base.cpp:
```diff
@@ -1,401 +1,1148 @@
+
+   static const auto bc_settings = bc::settings(bc::config::settings::mainnet);
+   const chain::block genesis = bc_settings.genesis_block;
+   BOOST_REQUIRE(instance.create(genesis));
```
In some tests, we use `genesis.hash()` to create checkpoints etc.

In all others, where the tests doesn't need the hash, I have remove the extraneous variable.

kulpreet commented on include/bitcoin/database/data_base.hpp:
```diff
@@ -110,6 +110,11 @@ class BCD_API data_base
 
+    // BLOCK ORGANIZER (confirm)
+    /// Confirm a block that is already in candidate index and parent
+    /// is already confirmed
```
Changed

kulpreet commented on include/bitcoin/database/data_base.hpp:
```diff
@@ -110,6 +110,11 @@ class BCD_API data_base
+    // BLOCK ORGANIZER (confirm)
+    /// Confirm a block that is already in candidate index and parent
+    /// is already confirmed
+    code confirm(const hash_digest& block_hash, const size_t height);
```
Rebased

kulpreet commented on test/data_base.cpp:
```diff
@@ -1,401 +1,1148 @@
+using namespace boost::system;
+using namespace boost::filesystem;
+
+static void
```
Fixed everywhere

kulpreet commented on test/data_base.cpp:
```diff
@@ -1,401 +1,1148 @@
+   store_block_transactions(instance, *block2_ptr, 1);
+   store_block_transactions(instance, *block3_ptr, 1);
+
+   const auto headers_push_ptr = std::make_shared<const header_const_ptr_list>(header_const_ptr_list{
```
Done.

kulpreet commented on test/data_base.cpp:
```diff
@@ -1,401 +1,1148 @@
+   block_const_ptr block2_ptr = std::make_shared<const message::block>(read_block(MAINNET_BLOCK2));
+   block_const_ptr block3_ptr = std::make_shared<const message::block>(read_block(MAINNET_BLOCK3));
+   const auto blocks_push_ptr = std::make_shared<const block_const_ptr_list>(block_const_ptr_list{
+           block1_ptr, block2_ptr, block3_ptr });
```
Fixed everywhere

kulpreet commented on test/data_base.cpp:
```diff
@@ -1,401 +1,1148 @@
+    }
+
+    bool pop_above(header_const_ptr_list_ptr headers,
+                   const config::checkpoint& fork_point)
```
Fixed everywhere.

kulpreet commented on test/data_base.cpp:
```diff
@@ -1,401 +1,1148 @@
+
+   // setup ends
+
+   auto result = instance.pop_above(out_headers, config::checkpoint(block1.hash(), 0));
```
Done

kulpreet commented:
New commits pushed addressing all the styling issues. Have configured my editor to avoid these errors again.

Also, added commit to confirm transactions of the block being confirmed.

Finally, rebased to use libbitcoin-system.

evoskuil commented on src/data_base.cpp:
```diff
@@ -324,6 +324,31 @@ code data_base::reorganize(const config::checkpoint& fork_point,
     return result ? error::success : error::operation_failed;
 }
 
+code data_base::confirm(const hash_digest& block_hash,
```
This approach is simple but least optimal. By populating a set and then using the transactions.confirm() overload that walks the set, there is an unnecessary memory bump (up to ~4mb for BTC with segwit) during confirmation. This and the additional iteration can be avoided by confirming in the population loop.

Much more importantly, the read of the entire tx is defeating the intended optimization. The objective here is to read the set of tx offsets and then confirm each transaction without reading the entire block. This requires confirming using `database::transaction_result` while avoiding transition through `chain::transaction`. This allows the tx to be confirmed without being fully populated. transaction_result is not fully-deferred, so this is also not entirely optimal, but is at least the correct abstraction.

For much greater optimization the tx confirmation loop can be parallelized by tx, but let's investigate that as a later optimization once this is solid.

kulpreet commented:
Updated styles and switched to confirming transactions without populating chain::transaction objects
